stablehlo/dialect/StablehloOps.td

```diff
@@ -1343,7 +1343,8 @@ def StableHLO_WhileOp: StableHLO_Op<"while", [
 }
 
 def StableHLO_AllGatherOp : StableHLO_Op<"all_gather",
-    [ConditionallySpeculatable, SameOperandsAndResultElementType] /*all_gather_c6*/> {
+    [ConditionallySpeculatable,
+     HLO_PairwiseSameOperandAndResultElementType, /*all_gather_c6*/]> {
```

Comment:
Do we want `SameOperandsAndResultElementType`? Should we allow all_gathers to operate on different element kinds?

Reply:
Edit: Looks like based on this test we do want to do this pairwise, mixed types are allowed.
 https://github.com/openxla/xla/blob/a02433bdc5e6dbe68446d6c95c4959b5e1715cf9/xla/service/hlo_verifier_test.cc#L2936

Reply:
The spec has `(C6) type(results...) = type(operands...)` -- does this imply pairwise or should we update that?

Reply:
Editedit: Looks like it does, optimization_barrier has `(C1) type(operand...) = type(result...)` as well.

Reply:
Leaving this comment to make sure that there are mixed element type tests for all_gather and all_to_all.